### DISCUSSION TITLE
Logs: Ability to select and copy multiple logs from Explore to the clipboard

### DISCUSSION THREAD

**Discussion by adeverteuil** [grafana/grafana#55849] [Observability]
**What would you like to be added**:

Selecting text in Explore logs is awkward. It often triggers UI behavior like showing the detailed labels and parsed data. Sometimes partial lines are selected if the cursor isn't aimed precisely.

I would like to easily select raw log lines and copy them to my clipboard.

**Why is this needed**:

Raw text is still an excellent format for sharing and storing debugging and troubleshooting information. Raw text is easier to share on Slack, save in a ticket internal comment, or send back to a customer as an email attachment. It is more efficient than a screenshot too. Sometimes, working with text in Vim, Emacs or Less (with regex search) is faster than the web UI. It's just easier to access the logs in Explore than it is to SSH into a VM or use kubectl logs. Often Explore is the only way for non-system-admins to access logs themselves, like technical support teams, but they still want to work with text and Unix-like command line tools.

Additionally:
* Toggle to reverse the display order so they are in the "natural" log file order.
* Toggle to copy text with or without the data source provided time stamp.
* Copy to Clipboard button that just grabs the entire query result.
* Easily select a subset of lines and copy it to clipboard.

**Comment by adeverteuil** [comment]
Now that there is a Query Inspector in Grafana Explore (v7.2), it is possible to work around this more easily by copying the JSON data to a text file and running it through `jq`.

    jq -r ".response.data.result[].values[][1]" logs.json | sort

https://grafana.com/docs/grafana/latest/explore/#query-inspector
https://grafana.com/docs/grafana/latest/whatsnew/whats-new-in-v7-2/#inspect-queries-in-explore

**Comment by ivanahuckova** [comment]
@adeverteuil We have added full [Query Inspector with Data Tab](https://github.com/grafana/grafana/pull/32005) to the Explore, so it is even easier now. Moreover, we are working on [implementing Download logs in txt](https://github.com/grafana/grafana/issues/28752). But fyi, we are keeping this issue in mind and we are considering improved "per-row-actions".

**Comment by iainlane** [comment]
I'm bumping up against this now. I want to copy 20 or so lines from a query into a new issue, but if I turn off time and c+p the rows I get results like

```
  |   | 1 file changed, 34 insertions(+)
```

the leading `  |   | ` is not from the logs themselves, it's being added by explore somehow. I can try the workaround mentioned but it feels like high friction.

**Comment by ivanahuckova** [comment]
+cc @svennergr not sure if we consider this closed by new "copy to clipboard" button

**Comment by iainlane** [comment]
FTR I don't see such a button. I'm running version v9.2.0-34885pre (8bc8778287).

**Comment by ivanahuckova** [comment]
There is a bug that button appears only when you use line filters. I have created a new issue for it: https://github.com/grafana/grafana/issues/55754

**Comment by iainlane** [comment]
Ah ok thanks. From your screencast there it looks like that lets you copy a single line at a time. If that's right, I'd leave this issue open for the ability to select and copy many lines.

**Comment by ivanahuckova** [comment]
👍 I updated the title to make it more clear that this is related to multiple lines

**Comment by svennergr** [comment]
@adeverteuil / @iainlane 
Out of curiosity, are the "Download Logs/CSV" buttons in the query inspector doing the thing you are aiming at? How can we improve those buttons?

**Comment by iainlane** [comment]
> Out of curiosity, are the "Download Logs/CSV" buttons in the query inspector doing the thing you are aiming at? How can we improve those buttons?

The main thing is that that flow is not really discoverable. I would expect to view the logs directly in Explore and be able by some obvious and quick (one click / key combo, ideally the normal copy ones) method be able to copy those to paste elsewhere. I just wouldn't think to go to a separate page to be able to copy logs out, it seems like something the main view should let me do.

A bit less important but still would be annoying for my usecase is that the "Download Logs" button gives me the timestamp in the file and I don't want to paste that so I'd have to do some manual work on the downloaded file to remove them.

**Comment by ivanahuckova** [comment]
As this is a feature request, I am moving it to discussions where we have all our feature requests.

**Comment by ivanahuckova** [comment]
For additional discussion, please use https://github.com/grafana/grafana/issues/77553
